### ISSUE TITLE
`prefer-const` does not detect variables declared outside loop

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#14900]
**What rule do you want to change?**

`prefer-const`

**Does this change cause the rule to produce more or fewer warnings?**

More warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default

**Please provide some example code that this change will affect:**

https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gZm9vICgpIHtcblx0bGV0IGE7XG5cdHdoaWxlIChNYXRoLnJhbmRvbSgpID4gMC41KSB7XG5cdFx0YSA9IGdldCgpO1xuXHRcdGNvbnNvbGUubG9nKGEpO1xuXHR9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJwcmVmZXItY29uc3QiOjJ9LCJlbnYiOnt9fX0=

```js
function foo () {
	let a;
	while (Math.random() > 0.5) {
		a = get();
		console.log(a);
	}
}
```
**What does the rule currently do for this code?**

Not detected

**What will the rule do after it's changed?**

`a` is only assigned and used in the loop, it should not be declared in the outer scope. There are only two situations where this usage is fine:

- `a` is used before assignment 

	
	```js
	function foo () {
		let a;
		while (Math.random() > 0.5) {
			a = get(a);
			console.log(a);
		}
	}
	```

- `a` is actually used in the outer scope

	
	```js
	function foo () {
		let a;
		while (Math.random() > 0.5) {
			a = get();
		}
		console.log(a);
	}
	```

**Comment by snitin315** [comment]
I believe this is intended behavior as this will also require us to change the position of declaration inside the loop, which seems out of scope for `prefer-const`.
 
```js
 function foo () {
 	while (Math.random() > 0.5) {
 		const a = get(a);
 		console.log(a);
 	}
 }
```

Also, both of these cases are mentioned in [**correct code** for this rule](https://eslint.org/docs/rules/prefer-const#rule-details) -

```js
// it's initialized in a different block from the declaration.
let a;
if (true) {
    a = 0;
}
console.log(a);

// it's initialized at a place that we cannot write a variable declaration.
let a;
if (true) a = 0;
console.log(a);

```

**Comment by fregante** [comment]
> Also, both of these cases are mentioned in [**correct code** for this rule](https://eslint.org/docs/rules/prefer-const#rule-details) -

Right, this issue is about making them _not_ correct.

> change the position of declaration inside the loop, which seems out of scope for `prefer-const`.

I don't see any difference between:

```js
let a;
a = 0;
```
and

```js
let a;
if (true) {
    a = 0;
}
```

other than "scope tracking" (I don't know the exact term), i.e. ensuring that `a` isn't used outside `if`

The diff for both of them is `s/let a//` and `s/a = 0/const a = 0/`

**Comment by snitin315** [comment]
This was intentional behavior before (https://github.com/eslint/eslint/issues/5251) but it seems as per https://github.com/eslint/eslint/issues/5251#issuecomment-186717772 the team decided to move this behavior to a new rule `prefer-smaller-scope`(https://github.com/eslint/eslint/issues/5284)

**Comment by snitin315** [comment]
Let's wait and see what other team members think about this proposal. 

/cc @eslint/eslint-team

**Comment by nzakas** [comment]
I don’t think prefer-const is the right place for this warning because you aren’t asking specifically if the variable can be a const, you’re asking if it’s declared in the right spot. I think the prefer-smaller-scope approach is the correct solution to this problem (assuming we still want to solve it)

@snitin315 please just move the issue into the Feedback Needed column if you are requesting help. The @-mention pings over 20 people including alumni, so we try not to do that. You can ping the TSC directly if you want more attention.

**Comment by btmills** [comment]
I agree that the `prefer-smaller-scope` approach is the right way to go here.

**Comment by ljharb** [comment]
A prefer-smaller-scope rule seems amazing, especially if it can be safely autofixable.

**Comment by mdjermanovic** [comment]
It seems that we have a consensus that this logic does not belong to `prefer-const`, so I'm closing this issue.

**Comment by fregante** [comment]
So is https://github.com/eslint/eslint/issues/5284 going to be reopened?

**Comment by mdjermanovic** [comment]
> So is #5284 going to be reopened? Or should I sent a new similar issue for a new rule (to avoid reopening an old issue)

Given the reason for closing the original issue https://github.com/eslint/eslint/issues/5284#issuecomment-272747601, and the description of another attempt to implement this rule https://github.com/eslint/eslint/pull/9265#issue-140145087, I'm not sure if can afford to maintain this rule in core ESLint, for the reasons described in [this blog post](https://eslint.org/blog/2020/05/changes-to-rules-policies).

@eslint/eslint-tsc should we reconsider adding `prefer-smaller-scope`? It feels that this rule would require significant maintenance effort, it doesn't relate to new ECMAScript features and it doesn't target possible errors, so I'm not in favor.

**Comment by brettz9** [comment]
`eslint-plugin-unicorn` seems to be carrying the torch for rules which might not be strictly necessary for preventing errors but which enforce good practices like this and help keep code clean. Maybe this could be proposed there.

**Comment by nzakas** [comment]
>  It feels that this rule would require significant maintenance effort, it doesn't relate to new ECMAScript features and it doesn't target possible errors, so I'm not in favor.

Agree.

**Comment by fregante** [comment]
For those interested, I opened an issue on unicorn: 

https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1524
